File: bazelrunner/src/main/java/org/jetbrains/bsp/bazel/bazelrunner/BazelRunnerBuildBuilder.java
package org.jetbrains.bsp.bazel.bazelrunner;

import ch.epfl.scala.bsp4j.BuildTargetIdentifier;
import com.google.common.collect.Lists;
import java.util.List;
import java.util.stream.Collectors;

public class BazelRunnerBuildBuilder extends BazelRunnerBuilder {

  private static final String EXCLUDABLE_TARGETS_LIST_PREFIX = "--";

  public BazelRunnerBuildBuilder(BazelRunner bazelRunner, String bazelBuildCommand) {
    super(bazelRunner, bazelBuildCommand);
  }

  @Override
  public BazelRunnerBuilder withTargets(List<String> bazelTargets) {
    return withArguments(bazelTargets);
  }

  @Override
  public BazelRunnerBuilder withTargets(
      List<BuildTargetIdentifier> includedTargets, List<BuildTargetIdentifier> excludedTargets) {
    var arguments = Lists.newArrayList(EXCLUDABLE_TARGETS_LIST_PREFIX);
    arguments.addAll(toRawUris(includedTargets));
    arguments.addAll(calculateExcludedTargetsWithExcludedPrefix(excludedTargets));

    return withArguments(arguments);
  }

  private List<String> toRawUris(List<BuildTargetIdentifier> targets) {
    return targets.stream().map(BuildTargetIdentifier::getUri).collect(Collectors.toList());
  }

  private List<String> calculateExcludedTargetsWithExcludedPrefix(
      List<BuildTargetIdentifier> targets) {
    return targets.stream()
        .map(BuildTargetIdentifier::getUri)
        .map(target -> "-" + target)
        .collect(Collectors.toList());
  }
}

File: bazelrunner/src/main/kotlin/org/jetbrains/bsp/bazel/bazelrunner/BazelRunnerBuildBuilder.kt
package org.jetbrains.bsp.bazel.bazelrunner

import ch.epfl.scala.bsp4j.BuildTargetIdentifier
import org.jetbrains.bsp.bazel.bazelrunner.utils.BazelArgumentsUtils
import org.jetbrains.bsp.bazel.workspacecontext.TargetsSpec

class BazelRunnerBuildBuilder(
    bazelRunner: BazelRunner,
    bazelBuildCommand: String
) : BazelRunnerBuilder(bazelRunner, bazelBuildCommand) {

  override fun withTargets(bazelTargets: List<String>): BazelRunnerBuilder {
    return withArguments(bazelTargets)
  }

  override fun withTargets(targetsSpec: TargetsSpec): BazelRunnerBuilder =
     withTargets(targetsSpec.values, targetsSpec.excludedValues)

  override fun withTargets(
      includedTargets: List<BuildTargetIdentifier>,
      excludedTargets: List<BuildTargetIdentifier>
  ): BazelRunnerBuilder {
    val arguments = mutableListOf(ExcludableTargetsListPrefix)
    arguments.addAll(BazelArgumentsUtils.toRawUris(includedTargets))
    arguments.addAll(
        BazelArgumentsUtils.calculateExcludedTargetsWithExcludedPrefix(excludedTargets))
    return withArguments(arguments)
  }

  companion object {
    private const val ExcludableTargetsListPrefix = "--"
  }
}
